rllib/algorithms/algorithm.py

```diff
@@ -1493,27 +1493,28 @@ def compute_single_action(
                 assert len(pp) == 1, "Only one preprocessor should be in the pipeline"
                 pp = pp[0]
 
-                # Note(Kourosh): The connector will leave the policy's connector in eval
-                # mode. would that be a problem?
-                pp.in_eval()
-                if observation is not None:
-                    _input_dict = {SampleBatch.OBS: observation}
-                elif input_dict is not None:
-                    _input_dict = {SampleBatch.OBS: input_dict[SampleBatch.OBS]}
-                else:
-                    raise ValueError(
-                        "Either observation or input_dict must be provided."
-                    )
+                if not pp.is_identity():
+                    # Note(Kourosh): This call will leave the policy's connector
+                    # in eval mode. would that be a problem?
+                    pp.in_eval()
+                    if observation is not None:
+                        _input_dict = {SampleBatch.OBS: observation}
+                    elif input_dict is not None:
+                        _input_dict = {SampleBatch.OBS: input_dict[SampleBatch.OBS]}
+                    else:
+                        raise ValueError(
+                            "Either observation or input_dict must be provided."
+                        )
 
-                # TODO (Kourosh): Create a new util method for algorithm that computes
-                # actions based on raw inputs from env and can keep track of its own
-                # internal state.
-                acd = AgentConnectorDataType("0", "0", _input_dict)
-                # make sure the state is reset since we are only applying the
-                # preprocessor
-                pp.reset(env_id="0")
-                ac_o = pp([acd])[0]
-                observation = ac_o.data[SampleBatch.OBS]
+                    # TODO (Kourosh): Create a new util method for algorithm that
+                    # computes actions based on raw inputs from env and can keep track
+                    # of its own internal state.
+                    acd = AgentConnectorDataType("0", "0", _input_dict)
+                    # make sure the state is reset since we are only applying the
+                    # preprocessor
+                    pp.reset(env_id="0")
```

Comment:
actually, why do we have such a large else clause for the connector case?
and why do we have all these pp logics special cased?
I feel like there is something with the way we do this ...

Reply:
I am simply trying to match the logic we had before. This is the high-level logic: 1) retrieve the preprocessor stack. 2) if not NonePrerpocessor apply it. All the corner cases here are because with connectors you can potentially chain multiple preprocessor connectors, and I wanted to explicitly error out so that we can fix it later if this becomes a use-case.

Reply:
huh, but then things will still be in a non-working state?
like I feel like the kind of branching you want is to branch away from this function entirely.
because you want agent connectors to handle all the transformations not just pp. and also we shouldn't do unsquash or clip actions below if connectors are enabled?